### DISCUSSION TITLE
SSH test using GitHub Action

### DISCUSSION THREAD

**Discussion by tdegeus** [community/community#24913] [Actions]
<p>I have a code that wraps around some SSH operations. I would like to build CI for this code. Is there a way to generate two virtual machine and SSH between them?</p>

**Comment by airtower-luna** [comment]
<p>Technically you should be able to use Qemu (possibly with libvirt) on an Ubuntu runner, maybe with Vagrant or something to set them up automatically.</p>
<p>However I’d recommend using containers instead if possible, because the performance is much better. I have a <a href="https://github.com/airtower-luna/autopkgtest-ssh-docker/blob/491fd9a5a6ac71d184c7422628186c2fd0351bb9/Dockerfile">Dockerfile that starts an SSH server</a> inside a container for a project of mine. In that project a script creates the necessary <code>authorized_keys</code> file after the container has been started, but you could also do that with a volume or in the Dockerfile. Maybe that example can help you get started. :slightly_smiling_face:</p>

**Comment by tdegeus** [comment]
<p>It’s actually a good idea to use an ssh server! I’ll try to get started based on your example, thanks.</p>

**Comment by tdegeus** [comment]
<p><a class="mention" href="/airtower-luna">@airtower-luna</a> I’ve been trying your suggestion. Here is what I have in my action:</p>
<pre><code class="lang-auto">    - name: Install SSH server
      if: runner.os == 'Linux'
      run: |
        sudo apt-get install openssh-server
        sudo systemctl enable ssh
        sudo systemctl start ssh
        mkdir -p "$HOME/.ssh"
        chmod 0700 "$HOME/.ssh"
        ssh-keygen -t rsa -q -f "$HOME/.ssh/id_rsa" -N ""
        cat "$HOME/.ssh/id_rsa.pub" &gt;&gt; "$HOME/.ssh/authorized_keys"
        chmod 640 "$HOME/.ssh/authorized_keys"
        echo "NoHostAuthenticationForLocalhost yes" &gt;&gt; ~/.ssh/config
        sudo systemctl restart ssh
        ssh localhost "uname"
</code></pre>
<p>However, I’m getting permission issues:</p>
<pre><code class="lang-auto">Reading package lists...
Building dependency tree...
Reading state information...
openssh-server is already the newest version (1:8.2p1-4ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 4 not upgraded.
Synchronizing state of ssh.service with SysV service script with /lib/systemd/systemd-sysv-install.
Executing: /lib/systemd/systemd-sysv-install enable ssh
Permission denied, please try again.
Permission denied, please try again.
runner@localhost: Permission denied (publickey,password).
</code></pre>
<p>would you happen to know how to solve these issues?</p>

**Comment by airtower-luna** [comment]
<p>At a glance, no. To get more information try running <code>ssh</code> with the <code>-v</code> option, and check <code>/var/log/auth.log</code> (as in, <code>cat</code> to the log or upload as artifact) for the server side.</p>

**Comment by tdegeus** [comment]
<p>Thanks <a class="mention" href="/airtower-luna">@airtower-luna</a> ! I’m not too good at reading the verbose output of ssh, but it seems to me that ssh wants to read a passphrase that I set to be empty:</p>
<pre><code class="lang-auto">debug1: Reading configuration data /etc/ssh/ssh_config
debug1: /etc/ssh/ssh_config line 19: include /etc/ssh/ssh_config.d/*.conf matched no files
debug1: /etc/ssh/ssh_config line 21: Applying options for *
debug1: Connecting to localhost [::1] port 22.
debug1: Connection established.
debug1: identity file /home/runner/.ssh/id_rsa type 0
debug1: identity file /home/runner/.ssh/id_rsa-cert type -1
debug1: identity file /home/runner/.ssh/id_dsa type -1
debug1: identity file /home/runner/.ssh/id_dsa-cert type -1
debug1: identity file /home/runner/.ssh/id_ecdsa type -1
debug1: identity file /home/runner/.ssh/id_ecdsa-cert type -1
debug1: identity file /home/runner/.ssh/id_ecdsa_sk type -1
debug1: identity file /home/runner/.ssh/id_ecdsa_sk-cert type -1
debug1: identity file /home/runner/.ssh/id_ed25519 type -1
debug1: identity file /home/runner/.ssh/id_ed25519-cert type -1
debug1: identity file /home/runner/.ssh/id_ed25519_sk type -1
debug1: identity file /home/runner/.ssh/id_ed25519_sk-cert type -1
debug1: identity file /home/runner/.ssh/id_xmss type -1
debug1: identity file /home/runner/.ssh/id_xmss-cert type -1
debug1: Local version string SSH-2.0-OpenSSH_8.2p1 Ubuntu-4ubuntu0.1
debug1: Remote protocol version 2.0, remote software version OpenSSH_8.2p1 Ubuntu-4ubuntu0.1
debug1: match: OpenSSH_8.2p1 Ubuntu-4ubuntu0.1 pat OpenSSH* compat 0x04000000
debug1: Authenticating to localhost:22 as 'runner'
debug1: SSH2_MSG_KEXINIT sent
debug1: SSH2_MSG_KEXINIT received
debug1: kex: algorithm: curve25519-sha256
debug1: kex: host key algorithm: ecdsa-sha2-nistp256
debug1: kex: server-&gt;client cipher: chacha20-poly1305@openssh.com MAC: &lt;implicit&gt; compression: none
debug1: kex: client-&gt;server cipher: chacha20-poly1305@openssh.com MAC: &lt;implicit&gt; compression: none
debug1: expecting SSH2_MSG_KEX_ECDH_REPLY
debug1: Server host key: ecdsa-sha2-nistp256 SHA256:NFnWPH2boToKCX0yq1K33LY0K3EwK810f599LpwoXJI
debug1: read_passphrase: can't open /dev/tty: No such device or address
Host key verification failed.
</code></pre>

**Comment by airtower-luna** [comment]
<p>This looks like the key issue, not the passphrase:</p>
<aside class="quote no-group" data-username="tdegeus" data-post="6" data-topic="166717">
<div class="title">
<div class="quote-controls"></div>
 tdegeus:</div>
<blockquote>
<pre><code class="lang-nohighlight">debug1: Server host key: ecdsa-sha2-nistp256 SHA256:NFnWPH2boToKCX0yq1K33LY0K3EwK810f599LpwoXJI
debug1: read_passphrase: can't open /dev/tty: No such device or address
Host key verification failed.
</code></pre>
</blockquote>
</aside>
<p>You need to either import the host key for localhost into <code>authorized_keys</code>, or (if you’re only going to make test connections to localhost, nothing critical) you can write <code>StrictHostKeyChecking accept-new</code> to automatically accept the new key.</p>

**Comment by tdegeus** [comment]
<p>Thanks once more <a class="mention" href="/airtower-luna">@airtower-luna</a> .</p>
<p>The thing that confuses me is that I did import the key:</p>
<pre><code class="lang-auto">ssh-keygen -t rsa -q -f "$HOME/.ssh/id_rsa" -N ""
cat "$HOME/.ssh/id_rsa.pub" &gt;&gt; "$HOME/.ssh/authorized_keys"
</code></pre>
<p>and that it did find it:</p>
<pre><code class="lang-auto">debug1: identity file /home/runner/.ssh/id_rsa type 0
</code></pre>

**Comment by airtower-luna** [comment]
<p>Sorry, writing <code>authorized_keys</code> in the previous post was a mistake. The issue is with <code>known_hosts</code>. The rest still applies, though. :wink:  The key missing from <code>known_hosts</code> is not your login key, but the <em>host</em> key, which the SSH client uses to ensure it’s talking to the right server.</p>

**Comment by tdegeus** [comment]
<p>Thanks! Indeed that was one issue in adding myself to <code>know_hosts</code>. However, even with this fixed, ssh seems to really now want to take my <code>rsa</code> key.</p>
<pre><code class="lang-auto">debug1: Offering public key: /home/runner/.ssh/id_rsa RSA SHA256:ivLPhVUx8LHGGeIQkIdDwBJVzp9wQBkRiFRkXXN5HwE agent
debug1: Authentications that can continue: publickey,password
debug1: Trying private key: /home/runner/.ssh/id_dsa
</code></pre>
<p>(Not to worry about publishing the key here, it is generated in a virtual machine in CI).</p>
<p>For reference:</p>
<aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/tdegeus/shelephant/pull/81/checks?check_run_id=2075412224" target="_blank" rel="noopener nofollow ugc">github.com/tdegeus/shelephant</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/tdegeus/shelephant/pull/81" target="_blank" rel="noopener nofollow ugc">SSH test on CI</a>
    </h4>

    <div class="branches">
      <code>tdegeus:main</code> ← <code>tdegeus:ci</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-03-09" data-time="18:17:57" data-timezone="UTC">06:17PM - 09 Mar 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/tdegeus" target="_blank" rel="noopener nofollow ugc">
          <img alt="tdegeus" src="https://user-images.githubusercontent.com/7030445/181087681-0b2d5e89-5171-464e-91a9-25f9d4fb11d8.jpeg" class="onebox-avatar-inline" width="20" height="20">
          tdegeus
        </a>
      </div>

      <div class="lines" title="17 commits changed 1 files with 39 additions and 7 deletions">
        <a href="https://github.com/tdegeus/shelephant/pull/81/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+39</span>
          <span class="removed">-7</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>
<aside class="onebox githubpullrequest">
  <header class="source">
      <a href="https://github.com/tdegeus/shelephant/pull/81" target="_blank" rel="noopener nofollow ugc">github.com/tdegeus/shelephant</a>
  </header>
  <article class="onebox-body">
    <div class="github-row">
  <div class="github-icon-container" title="Pull Request">
    <svg width="60" height="60" class="github-icon" viewbox="0 0 12 16" aria-hidden="true"><path d="M11 11.28V5c-.03-.78-.34-1.47-.94-2.06C9.46 2.35 8.78 2.03 8 2H7V0L4 3l3 3V4h1c.27.02.48.11.69.31.21.2.3.42.31.69v6.28A1.993 1.993 0 0 0 10 15a1.993 1.993 0 0 0 1-3.72zm-1 2.92c-.66 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2zM4 3c0-1.11-.89-2-2-2a1.993 1.993 0 0 0-1 3.72v6.56A1.993 1.993 0 0 0 2 15a1.993 1.993 0 0 0 1-3.72V4.72c.59-.34 1-.98 1-1.72zm-.8 10c0 .66-.55 1.2-1.2 1.2-.65 0-1.2-.55-1.2-1.2 0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2zM2 4.2C1.34 4.2.8 3.65.8 3c0-.65.55-1.2 1.2-1.2.65 0 1.2.55 1.2 1.2 0 .65-.55 1.2-1.2 1.2z"></path></svg>
  </div>

  <div class="github-info-container">
    <h4>
      <a href="https://github.com/tdegeus/shelephant/pull/81" target="_blank" rel="noopener nofollow ugc">SSH test on CI</a>
    </h4>

    <div class="branches">
      <code>tdegeus:main</code> ← <code>tdegeus:ci</code>
    </div>

    <div class="github-info">
      <div class="date">
        opened <span class="discourse-local-date" data-format="ll" data-date="2021-03-09" data-time="18:17:57" data-timezone="UTC">06:17PM - 09 Mar 21 UTC</span>
      </div>

      <div class="user">
        <a href="https://github.com/tdegeus" target="_blank" rel="noopener nofollow ugc">
          <img alt="tdegeus" src="https://user-images.githubusercontent.com/7030445/181087682-0a5021d2-22a3-4535-9f24-835402587797.jpeg" class="onebox-avatar-inline" width="20" height="20">
          tdegeus
        </a>
      </div>

      <div class="lines" title="17 commits changed 1 files with 39 additions and 7 deletions">
        <a href="https://github.com/tdegeus/shelephant/pull/81/files" target="_blank" rel="noopener nofollow ugc">
          <span class="added">+39</span>
          <span class="removed">-7</span>
        </a>
      </div>
    </div>

  </div>
</div>

  </article>
  <div class="onebox-metadata">
    
    
  </div>
  <div style="clear: both"></div>
</aside>

**Comment by airtower-luna** [comment, marked as answer]
<p>It looks like the SSH server rejected the key. To figure out why you’ll need to check <code>/var/log/auth.log</code> as I mentioned above, it should contain information on login attempts.</p>

**Comment by tdegeus** [comment]
<p>Thanks for all your help. The solution was in the end very simple. My $HOME has permissions 777, which is too much for SSH. So what solved it was</p>
<pre><code class="lang-auto">chmod g-w $HOME
chmod o-w $HOME
</code></pre>

**Comment by BradleyA** [comment]
<p>I have lost a lot of time debugging users ssh incidents only to find out it was a simple file permission.  I wrote a few bash scripts that allow users and administrators to make sure that the ssh files and directory permissions are correct on each system.   The script names are check-user-ssh.sh and check-host-ssh.sh which can be found at <a href="https://github.com/BradleyA/docker-security-infrastructure/tree/master/ssh/" class="inline-onebox" rel="noopener nofollow ugc">docker-security-infrastructure/ssh at master · BradleyA/docker-security-infrastructure · GitHub</a></p>

**Comment by tdegeus** [comment]
<p>Thanks nice <a class="mention" href="/bradleya">@bradleya</a> . Did you consider making this an GitHub Action?</p>
